### ISSUE TITLE
Integrating into Fellow Oak DICOM

### ISSUE THREAD

**Issue by amoerie** [Efferent-Health/fo-dicom.Codecs#6]
Hi! I'm a contributor to the [the Fellow Oak DICOM](https://github.com/fo-dicom/fo-dicom) project, a project I'm sure you've heard about. 😉

First of all, I would like to thank you for the great work you're doing with the codecs! They actually seem better than what fo-dicom itself provides. It's very impressive work. 👍

We are planning on moving Fellow Oak DICOM to .NET Standard 2.0, and in doing so we're looking at the codecs inside fo-dicom and how we can move to being cross platform. One of our options that we are strongly considering is getting rid of the codecs inside fo-dicom and adopting the codecs in this repository. Of course, we don't want to do this without involving you in the process, that's why I'm creating this issue.

Our options, as far as I can see, are the following:

**A: The nuget dependency**
We could add a NuGet dependency to "Efferent.Native" inside the "fo-dicom" NuGet package. That way, the two projects stay completely isolated, just like today. Efferent.Native just becomes the "officially recommended" codec implementation that resides in a different repository. Just like today, the two projects have to be deployed separately. This would require no effort on your part. You maintain complete control of the codecs and their release schedule.

**B: The git submodule**
We could add a git submodule reference to this repository inside the fo-dicom repository. This would enable us to build and deploy both fo-dicom and these codecs from within fo-dicom, and have them deployed on NuGet under the same umbrella. Again, this would require no effort on your part. You maintain complete control of the codecs and their release schedule.

**C: Full adoption**
We could integrate the source code of these codecs into the fo-dicom repository. You would become a direct contributor to fo-dicom and the codecs + the fo-dicom library could be built and deployed under the same umbrella, just like in option B. This option essentially replaces the current codecs implementation in fo-dicom with yours, and you would become a core fo-dicom contributor in the process. You would no longer have complete control of the codecs and their release schedule, because fo-dicom releases have to be coordinated with the other contributors.

Of course, much will depend on how you feel about this. That's why I would like to ask you the following question:

Which option would you prefer? Or do you have another option that I'm missing? 
I'd like to involve you in this process to figure out a solution that is best for everyone.

**Comment by jaime-olivares** [comment]
Hi @amoerie,
Thanks for reviewing our library. I have read the discussion at https://github.com/fo-dicom/fo-dicom/issues/870 
I would start this post with a big YES, we want to collaborate on this. Now let's try to work out the details.

The story starts as follows: We identified the necessity of a cross-platform Dicom library 3 years ago and tried to contact fo-dicom at some point but didn't see the enthusiasm we see now. Forking fo-dicom and trying to build it entirely is a painful process and discouraged us from going on for a short time, especially because we use Visual Studio Code.
Finally, we found that the Transcoder option allowed us to keep independence and still interoperate with fo-dicom with minimal impedance.
From a project management perspective, I think the entire Dicom library (comm+storage) is distinct from the Codec portion. Indeed it imports several other 3rd-party projects for processing image buffers. Moreover, this sub-project is bloated because of the different versions (desktop, net core, etc.) with similar but different C++/MC source code.
Based on all that, our purpose is to build a library with a unified source code, probably controlled by several #define directives. Also, we removed the requirement of Managed C++ code and demonstrated it was not really necessary.

There are several reasons that make us think both projects shall be kept separated:
- The knowledge area of both projects is different. The Codec project has a low relationship with Dicom and indeed can theoretically be used for other purposes.
- The issues are also disconnected: it is either a Dicom communication or storage issue, or a Codec issue.
- There is a low relationship between changes in both projects. One shouldn't affect the other.
- fo-dicom can become a pure C# project, leaving the Codec project dealing with C++ and cross-platform complexities.
- Both projects can be updated separately in nuget. There is not a real need for synchronizing deployments unless there is a breaking change (i.e. Major version raise in fo-dicom).

Now regarding the alternatives:
A. We still think this is the best choice, for the reasons exposed above. The Codec will have its own management rather than being messed with all the other Dicom assets (code, documentation, tests, issues, PR, artifacts).
B. The git submodule doesn't add too much isolation of the mentioned assets
C. Similar to option B.

The isolation proposal doesn't mean that we are not keen to collaborate in a close manner. We can give you guys any admin rights and even change the name of the project to something like "fo-dicom/fo-dicom-codecs", with the unique conditions to keep the isolation and advertise the name of our company as partner/sponsor/co-owner. The reason is that we have invested more than 2,000 paid man-hour on this: Internally, it hasn't been a volunteer project. Moreover, we still have plans to be the main contributors to this project.

Now going to the technical side, we think we can try to use our codecs exclusively for the whole fo-dicom project, either desktop, netcore or mobile. As you suggested, the PCL support may be deprecated because Microsoft already did. Other platforms can be compiled by handling directives, as I mentioned. .Net Standard 2.0 may cover all the cases.

Regarding the concern of supporting legacy users, like PCL or .net <4.7, they can still use old nuget packages and if they are really active, eventually they will decide to move to a supported platform. Maybe we will provide them the biggest reason.

Hope this helps. I look forward to your response.

PS. The tests are not really broken. It is a version-reference issue with our nuget package. We will fix that a.s.a.p.

**Comment by amoerie** [comment]
Hi @jaime-olivares

I am delighted to read your response, and I look forward to working together on this with you. 👍

So, if I understand you correctly, you would strongly prefer to keep the codecs in a separate repository and on a separate release schedule. Your arguments for doing so seem very valid, so I think @gofal (the current steward of the fo-dicom project) will agree to this too, but I don't speak for him so I'll await his response here. 

It's great that you're suggesting to start up a fo-dicom/fo-dicom-codecs repository, I was actually going to propose the exact same thing. This makes it very explicit that these codecs would be the official implementation. As far as I'm concerned, you can be the repository admins and have a banner for your company as big as you want. You're paying for it in man hours, like you mentioned.

@gofal What are your thoughts here? Do you find these plans agreeable?

In summary of what's been said up to now, I think the following tasks lie before us:

- Rename or somehow migrate this repository to fo-dicom/fo-dicom-codecs. Ensure everyone in this repo here is admin of the new repository.
- Add verification tests for the current codecs in fo-dicom. This might be tricky. These tests should allow us to switch to the new codecs afterwards while guaranteeing compatibility and no regressions. I am willing to take a stab at this when I have the time.
- In the .NET standard branch of @gofal, add a nuget dependency to this new package use them as the default codecs. Remove the old codecs. 
- Verify the new codecs can be loaded and used on each supported platform.

**Comment by gofal** [comment]
Great! Thank you guys! I am absolutely delighted about the plans I read here from you two. And i fell very comfortable with the thoughts of having separate repositories / nuget-packages / release-cycles etc.

Currently your nuget package is called "Efferent.Native". It does not have "fo-dicom" contained in the name, thats why I assume that many users do not find it that easy when using the nuget package manager. So I wanted to ask if you could add "fo-dicom" somewhere to the package name, but I also would have excpected that you also keep your companies name in the package name to make it really visible for every user. But of course its your decission how to name the project and how close you want to place your project to fo-dicom or how distinct you want to keep it.

I am only unsure about the direction of the dependency. Currently fo-dicom defines an interface and Dicom-native references fo-dicom and implements this interface. With the new major release of fo-dicom this interface will change, so this is a chance to rethink it and optimize it.
Is it still the best to have this direction of dependency, or should we for some reason change and you define some interface and then fo-dicom loads and calls it? This would move the points of contact between DICOM and codec from your project to fo-dicom.

**Comment by jaime-olivares** [comment]
Hi guys,
I have some progress regarding this topic:
- I have renamed the GitHub repo to fo-dicom.Codecs, as that's the hypothetical NuGet Package name.
- I have added both you guys as administrators.
- I created a branch https://github.com/Efferent-Health/fo-dicom.Codecs/tree/transition with all the preliminary changes related to the rebranding of this project, including the README.

Some comments:
- It is pending to create the NuGet Package and complete the update of the branch, and the following merge into master.
- There were some conflicts with the *Dicom.Imaging.Codec* namespace, so I rather used *Dicom.Imaging.NativeCodec*.
- It is not possible to decouple this project from fo-dicom but the inverse case is, as it has worked until now. The user will need to add this NuGet Package as an extra, as done with *fo-dicom-Drawing* for example.

**Comment by amoerie** [comment]
Great stuff Jaime, thanks for jumping on this so quickly. Your branch and the landing page are looking great. Would you be open to moving this page under the fo-dicom umbrella? So it would be github.com/fo-dicom/fo-dicom.Codecs. If not, I completely understand and we will leave it like this, but I think it would help in affirming that this library has the official stamp of approval by Fellow Oak DICOM. 💪 The choice is yours.

@gofal If they say yes, are you able to do this?

In the meantime we're working to get a stable .NET Standard branch, you can follow our progress here: https://github.com/fo-dicom/fo-dicom/commits/netstandard

I assume that you @jaime-olivares are in the best position to implement the consumption side of the codecs in fo-dicom itself, so I would like to ask if you can take a stab at branching off the netstandard-branch and consuming your newly created nuget package with the cross-platform codecs? As far as I'm concerned, you have a blank cheque to change any codec related APIs to make the codecs integration easier/faster/better/... so I would say, now's your chance! 😉

Feel free to remove/rename the old codecs if you need to, I'll work on a test suite for the old codecs in the meantime. When you're done, I'll look into migrating that test suite to the new codecs API to verify everything still works.

Then we can use the pull request page or your branch --> netstandard to discuss any changes you want to make. How does that sound?

**Comment by jaime-olivares** [comment]
Hi guys,
Regarding the first part, I think there won't be a problem with becoming fo-dicom/fo-dicom.Codecs. Not sure about the transference steps.
The second part is more complex. I would start analyzing: why it is not good as it is now? (this is: using a transcoder class/object) and what would be considered better?
As we relied on the isolation provided by the transcoder class, we are not very knowledgeable about the integration details.

**Comment by jaime-olivares** [comment]
Looking at the code, there is a high coupling between the codec library and fo-dicom. It uses at least a dozen of classes and interfaces. Countersense, fo-dicom doesn't need of a codec library for compiling: it loads it dynamically.

So we can start keeping things as they are, with a little twist:
- Remove the Codec implementation from the build of Dicom.Core
- Include fo-dicom.Codecs as a dependency, just for distribution issues
- Leave the discovery mechanism for loading Dicom.NativeCodecs.dll

Some caveats:
- We have never built a 32-bits library. That's something that would take a lot of time. Currently we only produce a single codec library for 64-bits.
- Our library is .net standard 2.0 while Dicom.Core is 1.3. This may present build issues. I suggest to raise to 2.0.

**Comment by amoerie** [comment]
Thank you for your thoughts. Let me address a few points:

- If you believe the current transcoder loading mechanism is fine, I can agree to that. I'd still want to move it a little in the direction of .NET Core dependency injection though, but that's not going to be a huge change.

- In the [netstandard](https://github.com/fo-dicom/fo-dicom/tree/netstandard) branch, everything is moved to .NET Standard 2.0 already 👍

- Concerning the 32/64 bit debate, I think having just 64 bit is fine now in 2020, because I can't think of a scenario where only 32 bit would work. Do you know about any, @gofal ?

**Comment by gofal** [comment]
Thanks, thats all great. Regarding the interface. Its fine if we keep things as they are. But now we do have a breaking change, so I didn't want to miss the opportunity of rethinking it. 

As amoerie already said, everything will be netstandard 2.0, so thats fine.

I still know some scearios, where still 32bit is needed. If you have applications that have to interact to some native windows dependencies. we had some integration with Nuance SpeechMagic for medical speech recognition via COM. This required to run the application in 32 bit. 
But I am very sure, this times will go away soon. 
Maybe I could then build our old codec project only for 32 bit and push a package named "desktop32-codec" for this very limited audience.

**Comment by gofal** [comment]
Currently we have an TranscoderManager class. this class is responsible for loading the IDicomCodec implementations. This was done, because the TranscoderManager did some magic like searching for loaded assemblies via reflection and so on. The same code had to work in various very different platform implementations. But now thats not necessary any more.

What do you think about that: There will be a singleton TranscoderManager. By default this includes the few transcoder we have in pure c# (encode rle, decode rle, decode jpeglossless). When adding the nuget package fo-dicom.codecs then it registers its implementation of the IDicomCodecs to this TranscoderManager, instead of completely replacing the whole TranscoderManager instance.

Or are you doing some specific things within your TranscoderManager implementation?

The advantage I see is that developers are more flexible which codec they want to use. Since version 4.0.1 fo-dicom supports custom transfersyntaxes (https://github.com/fo-dicom/fo-dicom/wiki/Supported-Transfer-Syntaxes). So a user could register a codec for any transfersyntax (say using ffmpeg a decoder for a Video-Transfersyntax and then decode the frames of the video...) and still use your codecs for the usual transfersyntaxes. Or the developer wants to use your codecs, but for a specific transfersyntax he has his own implementation.

Looking forward to your thoughts about this.

**Comment by jaime-olivares** [comment]
Guys, these will be our next steps based on your feedback:
- Keep working with 2.0/64-bits. Good news. 
- Keep the codec discovery approach.
- Remove the built-in codecs
- A small tweak for accumulating codecs rather than fully replacing them (need to handle collisions)

Regarding the "very basic" codecs like JPEG/RLE, I still haven't analyzed the differences but we have found that the fo-dicom implementation lacked handling of some deviations, like non-RGB color systems. We may need to keep all codecs together for that reason.

Questions:
- We were working with the master branch. Do you suggest us to work with the 'netstandard' branch? 
- Do you have plans to deprecate some platform or framework version?

**Comment by gofal** [comment]
We tried to fix all known bugs in master branch (version 4.0.*) to make a "final" stable version there and continue on the new netstandard code. This is currently in "netstandard" branch and will be version 5.0.
The step from 4.0 to 5.0 will have lot of breaking changes. instead of compiling several platform assemblies there will now be only on netstandard library etc. 
I have planned to continue on 4.0 only with bugfixes in future, but no more big enhancements.

**Comment by gofal** [comment]
The "netstandard" branch is heavily work in progress. So the interfaces may change. Once the interfaces are stable enought, it will be merged in development branch. Then it will be forked/used by the community.

But when we agreed on the interface of the TranscoderManager we of course will not change it before we have talked to you.

**Comment by jaime-olivares** [comment]
@gofal, thanks for the clarifications. I would suggest the following course of actions:
- For fo-dicom 4.0, let's keep things as they are, with exception of the rebranding to `Efferent-Health/fo-dicom.Codecs` and explicit use of a TranscoderManager for loading/overriding the current codec set. You will need to provide some advise in your README.
- We will wait for you guys to stabilize fo-dicom 5.0 and then we will incorporate the fo-dicom.Codecs as needed, with the capability of combining with other codecs, and pass the control to you as `fo-dicom/fo-dicom.Codecs`.

In the meantime, we still have to do a continuous refinement to the codecs, especially with respect to color (scheme, resolution, palettes, LUT, etc.)

Does that sound good?

**Comment by amoerie** [comment]
> * For fo-dicom 4.0, let's keep things as they are, with exception of the rebranding to `Efferent-Health/fo-dicom.Codecs` and explicit use of a TranscoderManager for loading/overriding the current codec set. You will need to provide some advise in your README.

I don't think there will be big - if any - changes to the TranscoderManager in any 4.x release going forward, that release lane is just going into maintenance mode. Even switching to these better codecs is something I actually wouldn't do for the 4.x releases. Best to keep changes here minimal, for optimal safety and stability. We want to be able to say "if 5.x is not working for you, please keep using 4.x for a while longer".

> * We will wait for you guys to stabilize fo-dicom 5.0 and then we will incorporate the fo-dicom.Codecs as needed, with the capability of combining with other codecs...`.

Thanks, please be patient for another while, there is still quite a lot of work to do. 

> * ...and pass the control to you as `fo-dicom/fo-dicom.Codecs`.

Just to be clear, the 'control' will very much stay in your hands 100%, it's just a little bit of name changing to hopefully make things easier for the community. 

> * What do you think about that: There will be a singleton TranscoderManager. By default this includes the few transcoder we have in pure c# (encode rle, decode rle, decode jpeglossless). When adding the nuget package fo-dicom.codecs then it registers its implementation of the IDicomCodecs to this TranscoderManager, instead of completely replacing the whole TranscoderManager instance.

I'm not entirely sure we even need a singleton TranscoderManager. If I understand correctly, we simply need a decoder and an encoder for each transfer syntax. That means we simply need to be able to register a IEncoder<TTransferSyntax> and IDecoder<TTransferSyntax> in the dependency injection container. (which would probably resolve to singletons indeed).

Any class that needs encoding/decoding can then simply request those via the constructor.

The only downside of this approach would be that you can only configure encoders/decoders on application startup, not like now where you "set" the implementation. Although, it would still be possible to register multiple encoders for the same transfer syntax, and then just dynamically resolve the one you want at runtime. So no functionality is lost.

**Comment by gofal** [comment]
The TranscoderManager would do some little work. 
- Instead of registering each IDicomCodec via dependencyInjection this class could search for types that implement the interface and load it
- The TranscoderManager keeps the mapping of Transfersyntax to IDicomCodec. With DI you would need to have many interfaces like IJpegEncoder, IJpegLosslessEncoder, IRLEEncoder, etc to register so that you later can retrieve the matching Codec.
- The TranscoderManager can also handle if a new codec is registered for a transfesyntax and replace the old one
- Some other services - like a StoreSCPService - can query the TranscoderManager for all available codecs to reject the transfersyntaxes that can not be decoded.

But I would register this TranscoderManager via DI, so that someone can implement its own TranscoderManager, that behaves different and for example allows several Codecs for the same Transfersyntax or has an other search strategy or whatever.

**Comment by gofal** [comment]
the new version 5.0 runs green unit tests. Of course all except those, that have encodings involved.
Before I continue to make changes in the very core classes, I would like to see all unittests green to see if I break something.

So what I see so far, the API of ICodec and ITranscodermanager will not change. 
There is one user request, to be able to render only one frame of a huge multiframe image without loading the whole file. Currently IDicomCodec always Encodes/Decodes all frames. Do you have an idea, how we could extend the interface so that only one frame is decoded instead the whole file?

Anyway: I've uploaded a alpha-version of the 5.0 here: https://www.nuget.org/packages/fo-dicom/5.0.0-alpha1
It would be great to get a nugetpackage fo-dicom.codecs 5.0.0-alpha1 that I can reference in unittests.
Will you do that, or should I create a new branch here and update the classes in \src\Codec?

**Comment by jaime-olivares** [comment]
Hi @gofal sorry for the late response. This covid19 situation has slowed us down. I will resume this project soon.

**Comment by amoerie** [comment]
Don't worry about it at all, I think COVID 19 is hurting us all. Health before work! I hope everyone can stay in good health and ride this pandemic out safely. 👍

**Comment by gofal** [comment]
Totally agree!
And covid changes so much now. Some are stuck at home and cant do nothing, and some others who deal with remote-working-software are now extremely busy because everyone needs remote access. I don'nt know anyone who says that its now just like before for him

**Comment by gofal** [comment]
I had some time, checked out the code trom branch "transition" and created a new project "fo-dicom.codecs.5.csproj" which imports the existing files, but makes them build with fo-dicom5.
Shall I commit?

**Comment by jaime-olivares** [comment]
Hi @gofal,
We are now back to work! 
Please let me know what you need and we will try to attend a.s.a.p.

**Comment by gofal** [comment]
Great to hear that news. I hope you all went through this healthy!

I took a look at the code in branch "transition" and saw the projects referencing fo-dicom 4.*. I then created new projects referencing fo-dicom 5.*. Thats because namespaces and interface changed. 

Shall I commit this into the branch?

I uploaded a nugetpackage fo-dicom 5.0.0.0-alpha1. This package is far not stable, but with that it is possible to reference to fo-dicom5. Now next what I need, is some build of fo-dicom.Codecs for fo-dicom5 - maybe also as alpha-package in nuget. Then I can integrate it in further development, unittests, ...
Currently  I am developing without all unittests where encoding/decoding is involved and therefore I cannot be sure if some changes break the handling of encoding/decoding. I really would get rid of that.

Thanks!

**Comment by jaime-olivares** [comment]
Hi @gofal,
We will need one more day to stabilize and test the big merge we did from the `transition` branch, as well as start publishing in nuget with the new branding. After that, you will be able to use it.
Regarding tests, you may see we have tests for every TS (results published in the release section). We will need to realign them with fo-dicom 5.
Will be in touch,
Jaime.

**Comment by jaime-olivares** [comment]
UPDATE: We finally published in nuget the new package: https://www.nuget.org/packages/fo-dicom.Codecs
It is still pointing to fo-dicom 4 but will redirect and test on Monday.

**Comment by jaime-olivares** [comment]
@gofal what is the version with should try against? We see the unique v5 package (5.0.0-alpha1) is marked as deprecated

**Comment by gofal** [comment]
yes, marked it as deprecated, because some users tried to use it and reported issues that some things are not working... 
Is this deprecated-label blocking? can you download it anyway? Otherwise I will upload an other verson alpha2

**Comment by jaime-olivares** [comment]
Let us try, probably not a blocker but just a warning

**Comment by jaime-olivares** [comment]
Hi guys, we have completed the integration testing of fo-dicom.Codecs with fo-dicom 5.0
The unit and acceptance tests are running Ok.

We have replaced the default transcoder manager in fo-dicom with ours, with the following code:         
````sharp
        public static void SetImplementation()
        {
            // Preparation of a new service provider and host
            var sc = new ServiceCollection();
            sc.AddFellowOakDicom();
            sc.AddTranscoderManager<NativeTranscoderManager>();   // This is the line where we apply our codecs
            var pr = sc.BuildServiceProvider();
            var h = new DefaultServiceProviderHost(pr);
            
            // Calling Setup.SetupDI by reflection, as it is internal
            var assembly = typeof(FellowOakDicom.DicomSetupBuilder).Assembly;
            var type = assembly.GetType("FellowOakDicom.Setup");
            var method = type.GetMethod("SetupDI", new []{typeof(IServiceProviderHost)});
            method.Invoke(null, new[]{h});
        }
````

As you can see, you need to replace only one line in your code for using NativeTranscoderManager.
We still need to evaluate the circularity of references between fo-dicom and fo-dicom.Codecs. After that, I think we are done.

**Comment by gofal** [comment]
Great, thanks!

the documentation of how to initialize DI in version 5 is not ready until now, but when using within ASP.Net you will call `.AddTranscoderManager<NativeTranscoderManager>()` in `ConfigureServices`
If using outsie asp.net then the DI is created behind the scenes on first access, but can be initialized that way:

      new DicomSetupBuilder()
          .RegisterServices(s => s.AddFellowOakDicom().AddTranscoderManager<NativeTranscoderManager>())
          // there will be more additional options like logging, defaultvalues, etc... e.g.
          .SkipValidation()
          .Build();

**Comment by jaime-olivares** [comment]
Got it!
We will change that in the tests, but that doesn't affect that fact that it is already tested.
You can start referencing it from fo-dicom as default transcoders for certain platforms.

**Comment by gofal** [comment]
I integrated the nugetpackage into the unittests and they work great!

There is only one test failing. When encoding PixelData into a lossy Jpeg like JpegBaseline, then the Photometric Interpretation has to be changed. See the issues here: 
https://github.com/fo-dicom/fo-dicom/issues/836
https://github.com/fo-dicom/fo-dicom/issues/921

Is that something that you will implement in your Codec?
If you do not update the DicomTags of the `PixelData`-class, but only the pixel data itself, then i will implement it in within the `DicomTranscoder` class.

**Comment by jaime-olivares** [comment]
Hi @gofal,
We have changed the photometric tag handling in -beta6
We think this handling should be done in the codec library rather than the fo-dicom library, as we do with other cases.
But be aware that this specific change you requested is causing issues in certain software (Radiant: not shown correctly, MicroDicom, shown correctly as they change back the photometric tag), while the previous version was shown correctly by Radiant and MicroDicom.

**Comment by gofal** [comment]
Many thanks!
So to understand right: now the photometric interpretation is changed as described in DICOM standard. But Radiant has a kind of bug and does not show these files correctly?

I think at the moment the integration is quite done. I was affraid there could be more issues, but it worked out of the box like a charm 😃 
I hope we will release the version 5 soon without a pre-release attribute, then you can also release the codec without the pre-release attribute in the version name.

**Comment by amoerie** [comment]
Hi guys, first of all thanks for the speedy progress while I was a little absent! Recently, much of time has been consumed entirely because I have recently become a father. 😀  Lately I have found a little time for fo-dicom again though. 

I hope you don't mind, I actually came to ask a question / report a problem. See my comment here please: https://github.com/fo-dicom/fo-dicom/pull/1058#issuecomment-657431149 

In the past, there was a unit test that called 100 transcode method calls on the same DICOM Dataset in parallel. Surprisingly, the old codecs seemed to be able to deal with that pretty well. Since we're using the new codecs now, this test will crash the test host occasionally. (Yes, it's THAT kind of problem)

@jaime-olivares Are you aware of any such limitation in the new codecs, and is this something we need to document? I have the impression that giving each method call a unique clone of the dataset avoids the problem, but I'm not 100% confident yet.

**Comment by jaime-olivares** [comment]
@amoerie Congrats for your new child!
We wukk try to find the root cause of the crash and be sure the code is re-entrant.
Can you point us to the tests you were running? I think those should be included in this project.

**Comment by amoerie** [comment]
@jaime-olivares Thanks! 

The test I'm referencing is this one: 

https://github.com/fo-dicom/fo-dicom/blob/f22da83962f255dd3f8369d6d3bf10a462e34e59/Tests/FO-DICOM.Tests/Bugs/GH549.cs#L19-L47

I think the intention of the original author was to create n copies of the DicomDataset, but in reality the "datasets" variable is a list that contains n references to the exact same DicomDataset. 

When I ran the tests on .NET Core with the following command:

`dotnet test ./Tests/FO-DICOM.Tests/FO-DICOM.Tests.csproj --configuration Release --framework netcoreapp2.1 --blame --diag diag.log`

I would get an error saying that the test host process has crashed, and that the last running test was "DicomTranscoderTranscode_ToCompressedCodecInParallel_NoMultithreadIssues"

Since modifying the code that it uses unique DICOM datasets (see this commit: https://github.com/fo-dicom/fo-dicom/pull/1058/commits/356b57a16d38e9ec69a44851fada18fb2ba1136e ) the process no longer seems to crash.

**Comment by amoerie** [comment]
Minor update: I created an issue here to track this from the fo-dicom side: https://github.com/fo-dicom/fo-dicom/issues/1072

**Comment by jaime-olivares** [comment]
We have released a new production version (5.0.0) in NuGet:
https://www.nuget.org/packages/fo-dicom.Codecs/5.0.0 

We sill expect some refinements. So, please consider it may be fixed with any new version as 5.x.y

**Comment by jaime-olivares** [comment]
Closing this topic. It seems the codecs are stable now.
